samcli/lib/bootstrap/companion_stack/companion_stack_manager.py

```diff
@@ -0,0 +1,259 @@
+        waiter = self._cfn_client.get_waiter("stack_delete_complete")
+        waiter_config = {"Delay": 10, "MaxAttempts": 60}
+        self._cfn_client.delete_stack(StackName=stack_name)
+        waiter.wait(StackName=stack_name, WaiterConfig=waiter_config)
+
+    def list_deployed_repos(self) -> List[ECRRepo]:
+        """
+        List deployed ECR repos for this companion stack
+        Not using create_change_set as it is slow.
+
+        Returns
+        -------
+        List[ECRRepo]
+            List of ECR repos deployed for this companion stack
+            Returns empty list if companion stack does not exist
+        """
+        if not self.does_companion_stack_exist():
+            return []
+        repos: List[ECRRepo] = list()
+        stack = boto3.resource("cloudformation", config=self._boto_config).Stack(self._companion_stack.stack_name)
+        resources = stack.resource_summaries.all()
+        for resource in resources:
+            if resource.resource_type == "AWS::ECR::Repository":
+                repos.append(
+                    ECRRepo(logical_id=resource.logical_resource_id, physical_id=resource.physical_resource_id)
+                )
+        return repos
+
+    def get_unreferenced_repos(self) -> List[ECRRepo]:
+        """
+        List deployed ECR repos that is not referenced by current list of functions
+
+        Returns
+        -------
+        List[ECRRepo]
+            List of deployed ECR repos that is not referenced by current list of functions
+            Returns empty list if companion stack does not exist
+        """
+        if not self.does_companion_stack_exist():
+            return []
+        deployed_repos: List[ECRRepo] = self.list_deployed_repos()
+        current_mapping = self._builder.repo_mapping
+
+        unreferenced_repos: List[ECRRepo] = list()
+        for deployed_repo in deployed_repos:
+            for _, current_repo in current_mapping.items():
+                if current_repo.logical_id == deployed_repo.logical_id:
+                    break
+            else:
+                unreferenced_repos.append(deployed_repo)
+        return unreferenced_repos
+
+    def delete_unreferenced_repos(self) -> None:
+        """
+        Blocking call to delete all deployed ECR repos that are unreferenced by a function
+        """
+        repos = self.get_unreferenced_repos()
+        for repo in repos:
+            try:
+                self._ecr_client.delete_repository(repositoryName=repo.physical_id, force=True)
+            except self._ecr_client.exceptions.RepositoryNotFoundException:
+                pass
+
+    def sync_repos(self) -> None:
+        """
+        Blocking call to sync companion stack with the following actions
+        Create/Update companion stack.
+        Deletes unreferenced repos.
+        Deletes companion stack if there isn't any repo left.
+        """
+        exists = self.does_companion_stack_exist()
```

Comment:
Q: what do we consider hard failure vs soft failure during sync of repos. Lets mention that in the docstring to document behavior.

Reply:
I'm not too sure on what you mean by soft failures in this case. The hard failures here are assume to be errors from API calls.